### PR TITLE
Upgrade keyvault sdk from T1 to T2

### PR DESCRIPTION
**Pull request by AFengKK**
1. Upgrade keyvault sdk from T1 to T2
2. Upgrade files:

  - key_vault_sample_base.py
  - backup_restore_sample.py
  - soft_delete_sample.py
   

@jameszliao-msft, @lmazuel, @AlexGhiondea, @chlowell for notification.

### CODE REVIEW

**Review by laiapat — COMMENTED** [review]
Should we also add a `backup_restore_certificate` sample using `azure-keyvault-certificates`'s [backup_certificate](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/_client.py#L483) and [restore_certificate_backup](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/keyvault/azure-keyvault-certificates/azure/keyvault/certificates/_client.py#L513) methods?

File: key_vault_sample_base.py [hunk 1]
```diff
@@ -124,21 +113,20 @@ def setup_sample(self):
...
-                                                          tenant=self.config.tenant_id)
-            self.resource_mgmt_client = ResourceManagementClient(self.mgmt_creds, self.config.subscription_id)
+            self.credential = DefaultAzureCredential()
 
-            # ensure the service principle has key vault as a valid provider
-            self.resource_mgmt_client.providers.register('Microsoft.KeyVault')
+            self.resource_mgmt_client = ResourceManagementClient(
+            credential=self.credential,
+            subscription_id=self.config.subscription_id
+            )
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply key_vault_sample_base.py]: I have updated by your comments.

File: soft_delete_sample.py [hunk 2]
```diff
@@ -36,40 +35,9 @@ def create_soft_delete_enabled_vault(self):
...
         print('creating soft delete enabled vault: {}'.format(vault_name))
 
-        # create the vault
-        vault = self.keyvault_mgmt_client.vaults.create_or_update(self.config.group_name, vault_name, parameters)
-
-        # wait for vault DNS entry to be created
-        # see issue: https://github.com/Azure/azure-sdk-for-python/issues/1172
-        self._poll_for_vault_connection(vault.properties.vault_uri)
-
+        vault = self.create_vault()
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py:40]: We do not need to update `self.create_vault()`, because creating a vault with soft-delete enabled by default.

**Review by laiapat — COMMENTED** [review]
Thank you for the updates! I did try running these locally and ran into permissions errors for the created vaults though (for `backup_restore_sample.py` and `soft_delete_sample.py`). I checked the vault in the Portal and saw that the correct access policy was created, so I'm not sure what the issue is. I added a `time.sleep(180)` after vault creation but still got the error. Do you know what might be happening?

Here's an example of an error message I got from running `soft_delete_sample.py`:
```
ERROR: running sample failed with raised exception:
HttpResponseError: (Forbidden) The user, group or application 'appid={client_id};oid={client_object_id};iss=https://sts.windows.net/{tenant_id}/' does not have keys create permission on key vault 'vault-bleak-jury-64985;location=westus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287
Code: Forbidden
Message: The user, group or application 'appid={client_id};oid={client_object_id};iss=https://sts.windows.net/{tenant_id}/' does not have keys create permission on key vault 'vault-bleak-jury-64985;location=westus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287
Inner error: {
    "code": "AccessDenied"
}
```

File: backup_restore_sample.py [hunk 3]
```diff
@@ -65,29 +82,85 @@ def backup_restore_key(self):
...
+        certificate_name = get_name('certificate')
+
+        certificate = first_certificate_client.begin_create_certificate(certificate_name, CertificatePolicy.get_default()).result()
+        print('created certificate {}'.format(certificate.name))
+
+        # list the certificates in the vault
+        certificate_properties = first_certificate_client.list_properties_of_certificates()
+        print("all of the certificates in the client's vault:")
+        for certificate_propertie in certificate_properties:
+            print(certificate_propertie.name)
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply backup_restore_sample.py]: I have updated by your comments.

File: backup_restore_sample.py [hunk 4]
```diff
@@ -65,29 +82,85 @@ def backup_restore_key(self):
...
+
+        # restore the certificate to the new vault
+        restored = second_certificate_client.restore_certificate_backup(backup)
+        print('restored certificate {}'.format(restored.name))
+
+        # list the certificates in the new vault
+        secret_properties = second_certificate_client.list_properties_of_certificates()
+        print("all of the certificates in the new vault:")
+        for secret_propertie in secret_properties:
+            print(secret_propertie.name)
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply backup_restore_sample.py]: I have updated by your comments.

File: soft_delete_sample.py [hunk 5]
```diff
@@ -36,49 +35,20 @@ def create_soft_delete_enabled_vault(self):
...
         print('vault {} created enable_soft_delete={}'.format(vault.name, vault.properties.enable_soft_delete))
 
 
     @keyvaultsample
     def enable_soft_delete_on_existing_vault(self):
         """        
-        enables soft delete on an existing vault
+        enables soft delete on an existing vault. 
+        Soft-delete will be enabled on all key vaults, the ability to opt out of soft-delete will be deprecated soon.
+        (See https://docs.microsoft.com/en-us/azure/key-vault/general/soft-delete-change)
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py]: I have updated by your comments.

File: soft_delete_sample.py [hunk 6]
```diff
@@ -36,49 +35,20 @@ def create_soft_delete_enabled_vault(self):
...
         """
-        # create a vault without soft delete enabled
+        # create a vault with soft delete enabled
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py:52]: I have updated by your comments.

File: soft_delete_sample.py [hunk 7]
```diff
@@ -114,49 +74,47 @@ def deleted_vault_recovery(self):
...
         # list the deleted vaults
         deleted_vaults = self.keyvault_mgmt_client.vaults.list_deleted()
-        print('Deleted Vaults: \n{}'.format(self._serialize(deleted_vaults)))
+        for deleted_vault in deleted_vaults:
+            print(deleted_vault.name)
 
         # get the details of a specific deleted vault
         deleted_info = self.keyvault_mgmt_client.vaults.get_deleted(vault_to_recover.name, vault_to_recover.location)
-        print('Deleted vault info for vault: {}\n{}'.format(vault_to_recover.name, deleted_info))
+        print('Deleted vault: {}'.format(deleted_info.properties))
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py]: I have updated by your comments. I updated to print out the vault's name here.

File: soft_delete_sample.py [hunk 8]
```diff
@@ -114,49 +74,47 @@ def deleted_vault_recovery(self):
...
         # list the deleted vaults again only the vault we intend to purge is still deleted
         deleted_vaults = self.keyvault_mgmt_client.vaults.list_deleted()
-        print('Deleted Vaults: \n{}'.format(self._serialize(deleted_vaults)))
+        for deleted_vault in deleted_vaults:
+                    print(deleted_vault.name)
 
         # purge the last deleted vault
-        self.keyvault_mgmt_client.vaults.purge_deleted(vault_to_purge.name, vault_to_purge.location)
-        print('Purged vault: {}'.format(vault_to_recover.name))
+        self.keyvault_mgmt_client.vaults.begin_purge_deleted(vault_to_purge.name, vault_to_purge.location)
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py]: I have updated by your comments.

File: soft_delete_sample.py [hunk 9]
```diff
@@ -114,49 +74,47 @@ def deleted_vault_recovery(self):
...
+        print('Purged vault: {}'.format(vault_to_purge.name))
 
         # verify no deleted vaults remain
         deleted_vaults = self.keyvault_mgmt_client.vaults.list_deleted()
-        print('Deleted Vaults: \n{}'.format(self._serialize(deleted_vaults)))
+        for deleted_vault in deleted_vaults:
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py:118]: I have updated by your comments.

File: soft_delete_sample.py [hunk 10]
```diff
@@ -215,47 +186,56 @@ def deleted_key_recovery(self):
...
+        for deleted_key in deleted_keys:
+            print(deleted_key.name)
 
         # recover a deleted key
-        key = self.keyvault_data_client.recover_deleted_key(vault.properties.vault_uri, key_to_recover)
-        self._wait_on_recover_completed(vault.properties.vault_uri, 'key', key_to_recover)
-        print('recovered key {}\n{}'.format(key_to_recover, self._serialize(key)))
+        recover_key_poller = key_client.begin_recover_deleted_key(key_to_recover)
+        recovered_key = recover_key_poller.result()
+        recover_key_poller.wait()
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py]: I have updated by your comments.

File: soft_delete_sample.py [hunk 11]
```diff
@@ -264,86 +244,59 @@ def deleted_certificate_recovery(self):
...
+        certificate_client = CertificateClient(vault_url=vault.properties.vault_uri, credential=credential)
 
         # create certificates in the vault
         cert_to_recover = get_name('cert')
         cert_to_purge = get_name('cert')
 
-        cert = self.keyvault_data_client.create_certificate(vault.properties.vault_uri, cert_to_recover)
-        print('created certificate {}\n{}'.format(cert_to_recover, self._serialize(cert)))
+        create_certificate_poller = certificate_client.begin_create_certificate(cert_to_recover, policy=CertificatePolicy.get_default())
+        print('created certificate {}'.format(create_certificate_poller.result()))
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py]: I have updated by your comments.

File: soft_delete_sample.py [hunk 12]
```diff
@@ -264,86 +244,59 @@ def deleted_certificate_recovery(self):
...
+        deleted_certificate_poller  = certificate_client.begin_delete_certificate(cert_to_purge)
+        deleted_certificate = deleted_certificate_poller.result()
+        deleted_certificate_poller.wait()
+        print('deleted certificate {}'.format(deleted_certificate.name))
 
         # list the deleted certificates
-        deleted_certs = self.keyvault_data_client.get_deleted_certificates(vault.properties.vault_uri)
-        print('deleted certificates: \n{}'.format(self._serialize(deleted_certs)))
+        deleted_certs = certificate_client.list_deleted_certificates()
+        print('list the deleted certificates')
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py]: I have updated by your comments.

File: soft_delete_sample.py [hunk 13]
```diff
@@ -264,86 +244,59 @@ def deleted_certificate_recovery(self):
...
-        self._wait_on_recover_completed(vault.properties.vault_uri, 'key', cert_to_recover)
-        print('recovered certificate {}\n{}'.format(cert_to_recover, self._serialize(cert)))
+        recovered_certificate_poller = certificate_client.begin_recover_deleted_certificate(cert_to_recover)
+        recovered_certificate_certificate = recovered_certificate_poller.result()
+        recovered_certificate_poller.wait()
+        print('recovered certificate {}'.format(recovered_certificate_certificate.name))
 
         # purge a deleted certificate
-        self.keyvault_data_client.purge_deleted_certificate(vault.properties.vault_uri, cert_to_purge)
+        certificate_client.purge_deleted_certificate(cert_to_purge)
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py:295]: I have updated by your comments. I added `time.sleep(30)`s (here and on lines 174 and 232)

File: soft_delete_sample.py [hunk 14]
```diff
@@ -165,47 +123,60 @@ def deleted_secret_recovery(self):
...
+        for secret_property in secret_properties:
+            print(secret_property.name)
 
         # delete the secrets
-        deleted_secret = self.keyvault_data_client.delete_secret(vault.properties.vault_uri, secret_to_recover)
-        self._wait_on_delete_completed(vault.properties.vault_uri, 'secret', secret_to_recover)
-        print('deleted secret {}\n{}'.format(secret_to_recover, self._serialize(deleted_secret)))
+        delete_secret_poller = secret_client.begin_delete_secret(secret_to_recover)
+        deleted_secret = delete_secret_poller.result()
+        delete_secret_poller.wait()
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py:153]: I deleted `wait()`s (here and on line 158), and then an error was reported when running: 
```
azure.core.exceptions.ResourceExistsError: (Conflict) Secret secret-vague-show-20533 is currently being deleted.
Code: Conflict
Message: Secret secret-vague-show-20533 is currently being deleted.
Inner error: {
    "code": "ObjectIsBeingDeleted"
}
```

File: soft_delete_sample.py [hunk 15]
```diff
@@ -264,86 +244,59 @@ def deleted_certificate_recovery(self):
...
+        for certificate in certificates:
+            print(certificate.name)
 
         # delete the certificates
-        deleted_cert = self.keyvault_data_client.delete_certificate(vault.properties.vault_uri, cert_to_recover)
-        self._wait_on_delete_completed(vault.properties.vault_uri, 'certificate', cert_to_recover)
-        print('deleted certificate {}\n{}'.format(cert_to_recover, self._serialize(deleted_cert)))
+        deleted_certificate_poller  = certificate_client.begin_delete_certificate(cert_to_recover)
+        deleted_certificate = deleted_certificate_poller.result()
+        deleted_certificate_poller.wait()
```
[the remark this answers is not in the document]
Comment (AFengKK) [reply soft_delete_sample.py:275]: I deleted `wait()`s (here and on line 280), and then an error was reported when running: 
```
azure.core.exceptions.ResourceExistsError: (Conflict) Certificate cert-vain-bar-07455 is currently being deleted.
Code: Conflict
Message: Certificate cert-vain-bar-07455 is currently being deleted.
Inner error: {
    "code": "ObjectIsBeingDeleted"
}
```

**Comment by AFengKK** [comment]
> Thank you for the updates! I did try running these locally and ran into permissions errors for the created vaults though (for `backup_restore_sample.py` and `soft_delete_sample.py`). I checked the vault in the Portal and saw that the correct access policy was created, so I'm not sure what the issue is. I added a `time.sleep(180)` after vault creation but still got the error. Do you know what might be happening?
> 
> Here's an example of an error message I got from running `soft_delete_sample.py`:
> 
> ```
> ERROR: running sample failed with raised exception:
> HttpResponseError: (Forbidden) The user, group or application 'appid={client_id};oid={client_object_id};iss=https://sts.windows.net/{tenant_id}/' does not have keys create permission on key vault 'vault-bleak-jury-64985;location=westus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287
> Code: Forbidden
> Message: The user, group or application 'appid={client_id};oid={client_object_id};iss=https://sts.windows.net/{tenant_id}/' does not have keys create permission on key vault 'vault-bleak-jury-64985;location=westus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287
> Inner error: {
>     "code": "AccessDenied"
> }
> ```

I seem to have encountered such a problem at first, but I can’t remember. It may be that the environment variable `self.client_oid` in 'key_vault_sample_config.py' are configured incorrectly.

**Comment by AFengKK** [comment]
@mccoyp
The 'client_oid' is not the 'Object ID' in App, and it should be the 'oid' in the error message. You can try the "oid" value in the error message. If there are still errors, please contact me.
image

image

**Comment by mccoyp** [comment]
@AFengKK thank you so much for the info! I just tried using the OID from the error message and you're right, the samples are working now. I didn't realize there were different object IDs for my test application's App Registration and Enterprise Application. I'll run all the samples to verify that they work as expected and submit a review today.

File: soft_delete_sample.py [hunk 16]
```diff
@@ -114,49 +76,48 @@ def deleted_vault_recovery(self):
...
         print('Deleted vault: {}'.format(vault_to_recover.name))
-
         self.keyvault_mgmt_client.vaults.delete(self.config.group_name, vault_to_purge.name)
-        self._wait_on_delete_completed(None, 'vault', vault_to_purge.name)
         print('Deleted vault: {}'.format(vault_to_purge.name))
 
         # list the deleted vaults
         deleted_vaults = self.keyvault_mgmt_client.vaults.list_deleted()
-        print('Deleted Vaults: \n{}'.format(self._serialize(deleted_vaults)))
+        for deleted_vault in deleted_vaults:
```
Comment (mccoyp) [inline soft_delete_sample.py:90]: We should probably still print "Deleted vaults:" here and later in the sample to explain what the list of vaults is for:
```suggestion
        print("Deleted vaults:")
        for deleted_vault in deleted_vaults:
```

File: soft_delete_sample.py [hunk 17]
```diff
@@ -114,49 +76,48 @@ def deleted_vault_recovery(self):
...
+        recovery_parameters = VaultCreateOrUpdateParameters(location=deleted_vault.properties.location,
                                                             properties=recovery_properties)
-        recovered = self.keyvault_mgmt_client.vaults.create_or_update(self.config.group_name, deleted_info.name, recovery_parameters).result()
+        recovered = self.keyvault_mgmt_client.vaults.begin_create_or_update(self.config.group_name, deleted_vault.name, recovery_parameters).result()
         print('Recovered vault: {}'.format(recovered.name))
 
         # list the deleted vaults again only the vault we intend to purge is still deleted
         deleted_vaults = self.keyvault_mgmt_client.vaults.list_deleted()
-        print('Deleted Vaults: \n{}'.format(self._serialize(deleted_vaults)))
+        for deleted_vault in deleted_vaults:
```
Comment (mccoyp) [inline soft_delete_sample.py:108]:
```suggestion
        print("Deleted vaults:")
        for deleted_vault in deleted_vaults:
```

File: soft_delete_sample.py [hunk 18]
```diff
@@ -264,86 +246,61 @@ def deleted_certificate_recovery(self):
...
-        for x in range(max_retries):
-            try:
-                return func(*args[0:]) 
-            except KeyVaultErrorException as e:
-                if not e.response.status_code == 404:
-                    raise e
-                time.sleep(retry_wait)
+        # list the vault certificates
+        certificates = certificate_client.list_properties_of_certificates()
+        print('list the vault certificates')
```
Comment (mccoyp) [inline soft_delete_sample.py:301]: To align with print statements from the key and secret samples:
```suggestion
        print('all of the certificates in the client's vault:')
```

File: soft_delete_sample.py [hunk 19]
```diff
@@ -215,47 +188,56 @@ def deleted_key_recovery(self):
...
-        self.keyvault_data_client.purge_deleted_key(vault.properties.vault_uri, key_to_purge)
+        key_client.purge_deleted_key(key_to_purge)
+        time.sleep(50)
         print('purged key {}'.format(key_to_purge))
 
         # list the vaults key
-        keys = self.keyvault_data_client.get_keys(vault.properties.vault_uri)
-        print('keys: \n{}'.format(self._serialize(keys)))
+        keys = key_client.list_properties_of_keys()
+        print("keys:")
```
Comment (mccoyp) [inline soft_delete_sample.py:237]: To align with print statements from the certificate and secret samples:
```suggestion
        print("all of the keys in the client's vault:")
```

**Review by laiapat — APPROVED** [review]
Everything looks good -- just a few comments for output consistency. Thanks for all the work on this!

**Comment by AFengKK** [comment]
@mccoyp
I have updated by your comments.

**Review by laiapat — APPROVED** [review]

**Pull request merged**